support/scripts/multiboot2.py

```diff
@@ -0,0 +1,245 @@
+MULTIBOOT_HEADER_TAG_RELOCATABLE = 10
+
+
+class Elf64Editor:
+    endian = "<"
+
+    def __init__(self, elf64):
+        self.elf64 = elf64
+        self.file_sz = os.path.getsize(elf64)
+
+        with open(elf64, "rb") as f:
+            self.elf64_ehdr = f.read(ELF64_EHDR_LEN)
+
+        # EI_CLASS must be 2 for 64-bit
+        if self.elf64_ehdr[4] != 2:
+            raise Exception("File format is not ELF64")
+
+        # Check EI_DATA for endianness
+        if self.elf64_ehdr[5] == 1:
+            self.endian = "little"
+        else:
+            self.endian = "big"
+        _endian = ">" if self.endian == "big" else "<"
+
+        # Check e_machine
+        if sum(self.elf64_ehdr[18:20]) != ELF_MACHINE["x86_64"]:
+            raise Exception("Expected x86_64 ELF64. Wrong image format.")
+
+        # Gather whatever is needed in order to parse the Program Headers and
+        # the Section Headers. No need to store the original e_phentsize
+        # or e_shentsize since we will only ever use them here. Other than that,
+        # the file offsets should stay exactly the same.
+        self.e_phoff = unpack(_endian + "Q", self.elf64_ehdr[32:40])[0]
+        self.e_shoff = unpack(_endian + "Q", self.elf64_ehdr[40:48])[0]
+        ehdr64_e_phentsize = unpack(_endian + "h", self.elf64_ehdr[54:56])[0]
+        self.e_phnum = unpack(_endian + "H", self.elf64_ehdr[56:58])[0]
+        ehdr64_e_shentsize = unpack(_endian + "h", self.elf64_ehdr[58:60])[0]
+        self.e_shnum = unpack(_endian + "H", self.elf64_ehdr[60:62])[0]
+
+        with open(elf64, "rb") as f:
+            f.seek(self.e_phoff)
+            self.elf64_phdrs = []
+            for _ in range(self.e_phnum):
+                self.elf64_phdrs.append(f.read(ehdr64_e_phentsize))
+
+            f.seek(self.e_shoff)
+            self.elf64_shdrs = []
+            for _ in range(self.e_shnum):
+                self.elf64_shdrs.append(f.read(ehdr64_e_shentsize))
+
+    # We use this function to adjust offsets w.r.t. the prepended ELF64 header,
+    # the Multiboot header and the Program headers.
+    def prpnd_off(self, barray, mb2_header_end_off):
+        sz = len(barray)
+        old_off = int.from_bytes(barray, self.endian)
+
+        new_off = mb2_header_end_off + ELF64_PHDR_LEN * self.e_phnum + old_off
+        if sz < (new_off.bit_length() + 7) // 8:
+            raise Exception("New size exceeds initial byte-length.")
+
+        return new_off.to_bytes(sz, self.endian)
+
+    def get_ehdr(self, mb2_header_end_off):
+        elf64_ehdr = bytearray(ELF64_EHDR_LEN)
+        elf64_ehdr[0:ELF64_EHDR_LEN] = self.elf64_ehdr[0:ELF64_EHDR_LEN]
+
+        # The offset to the ELF64 Program/Sections Headers are at
+        # the end of the Multiboot Header and the end of the file respectively,
+        # so that we do not mess up the Multiboot2 header, which must exist
+        # in the first 8192 bytes. Although the specification does not enforce
```

Comment:
Perhaps mention that the spec mandates that it is in the first 32768 bytes though. This seems to be a GRUB quirk only.